File: dway_heap.py
'''
Created on 08/apr/2013

@author: mlarocca
'''


class dway_heap(object):
  
  PRIORITY_INDEX = 0
  ELEMENT_INDEX = 1
  
  def __init__(self, d, max_heap=False):
    ''' : Constructor :
        Create a new d-way heap priority queue.
        : param d : The branch-factor of the heap.
                    (MUST be at least 2)
        : type d : int (Other types will be converted to int)
        : param max_heap : If passed and truthy, the heap will 
                            be a max heap instead than a min heap.
        : type max_heap : boolean 
        : raise IllegalArgumentException : If d is less than 2.
        : return : self, as all constructors.
    '''
    self.__d = int(d)
    
    if self.__d < 2:
      raise Exception("IllegalArgumentException: minimum allowed branch-factor is 2")
    
    #if it is a max queue, uses the reverse ordering
    #The function projects the 
    self.__smaller = ((lambda x, y: x[dway_heap.PRIORITY_INDEX] > 
                                    y[dway_heap.PRIORITY_INDEX]) 
                      if max_heap 
                      else (lambda x, y: x[dway_heap.PRIORITY_INDEX] < 
                                         y[dway_heap.PRIORITY_INDEX])
                     )
    #same function, but without the need to project the priority the tuple
    self.__smaller_priority = ((lambda x, y: x > y) if max_heap 
                                else (lambda x, y: x < y))
    
    self.__queue = []

  def empty(self):
    ''' Shortcut for heap.size() == 0
        : return : True <=> there is no element in the heap
    '''    
    return len(self.__queue) <= 0
  
  def size(self):
    ''' Return the number of elements in the heap
        : return : The number of elements in the heap
    '''
    return len(self.__queue)

  def put(self, element, priority):
    ''' Insert a new element in the heap
        : param element : The element to insert.
        : param priority : The priority associated with the element.
        : return : self, to allow method chaining.
        
    '''
    new_item = (priority, element)
    self.__queue.append(new_item) #insert a placeholder to enlarge the array
    
    pos = len(self.__queue) - 1
    parent = (pos - 1) / self.__d
    
    while parent >= 0:
      
      if self.__smaller(new_item, self.__queue[parent]):
        
        self.__queue[pos] = self.__queue[parent]
        pos = parent
        parent = (parent - 1) / self.__d        
      else:
        parent = pos
        break
      
    if parent < 0:
      parent = 0
      
    self.__queue[parent] = new_item     
    assert self.check()
    
    return self
  
  def top(self):
    ''' Returns the top element, and removes it from the heap.
        
        : raise Exception : If the heap is empty.
        : return : The top element.
    '''
    size = self.size()
    if size == 0:
      raise Exception("The Heap is empty")
    
    #removes the last element from the queue
    item = self.__queue.pop()
    size -= 1
    
    if size == 0:
      return item[dway_heap.ELEMENT_INDEX]
    else:
      res = self.__queue[0]
      
      pos = 0
      tmp_pos = pos * self.__d + 1
      child_pos = tmp_pos  
          
      while child_pos < size:
        #Look for the smallest children
        i = 1
        while i < self.__d and tmp_pos + i < size:
          if self.__smaller(self.__queue[tmp_pos + i], self.__queue[child_pos]):
            child_pos = tmp_pos + i 
          i += 1
        if self.__smaller(self.__queue[child_pos], item):
          self.__queue[pos] = self.__queue[child_pos]
          
          pos = child_pos
          tmp_pos = pos * self.__d + 1
          child_pos = tmp_pos
        else:
          break
        
      self.__queue[pos] = item
      
      assert self.check()
      return res[dway_heap.ELEMENT_INDEX]
 
  def decrease_priority(self, element, priority):
    ''' Decrease the priority of a given key
        WARNING: duplicates keys aren't handled!
        : param element : The "key", aka the element whose priority must be 
                          decreased.
        : param priority : The new priority for the element. 
        : raise IllegalArgumentException : If the new priority for a key is 
                                           greater than the the existing one.
        : return : True iff the key was in the heap and its priority 
                   has been successfully updated.
    '''    
    for pos in xrange(self.size()):
      item = self.__queue[pos]
      if item[dway_heap.ELEMENT_INDEX] == element:
        if not self.__smaller_priority(priority, item[dway_heap.PRIORITY_INDEX]):
          raise Exception("Existing key priority can only be decreased!") 
        self.__queue[pos] = new_item = (priority, element)
        break
    if pos == self.size():
      return False  #Key not found
    
    parent = (pos - 1) / self.__d
    
    while parent >= 0:
      
      if self.__smaller(new_item, self.__queue[parent]):
        
        self.__queue[pos] = self.__queue[parent]
        pos = parent
        parent = (parent - 1) / self.__d        
      else:
        parent = pos
        break
      
    if parent < 0:
      parent = 0
    self.__queue[parent] = new_item 
    
    assert self.check()
    
  def check(self):
    ''' Check queue integrity
        : raise AssertionError : If the main property of the dway heap is violated
        : return : True iff the heap is valid
    '''
    pos = 0
    child = 1
    
    while child < self.size():
      i = 0
      while i < self.__d and child + i < self.size():
        assert not self.__smaller(self.__queue[child + i], self.__queue[pos])
        i += 1
        
      pos += 1
      child = pos * self.__d + 1
    
    return True
    
  def clear(self):
    ''' Remove all the elements in the heap.
        : return : self, to allow method chaining.
    '''
    self.__queue = []
    return self
    
  def heapsort(self):
    ''' Return a sorted array with all the elements in the heap.
        WARNING: All the elemens will be removed from the heap!
        
        : return : An array with the elements in the heap.
    '''
    if self.empty():
      return []
    else:
      res = []
      while not self.empty():
        res.append( self.top())
      
      return res
    
    
  def __str__(self):
    ''' : override :
    '''
    return " ".join(map(lambda x : str(x[dway_heap.ELEMENT_INDEX]), self.__queue))

    
if __name__ == '__main__':
  
  from random import randrange
  
  def test():
    ''' Test the data structure
    '''
    print "Test min heap"
    
    for d in xrange(2,10):
      #Test d way
      print d
      pq = dway_heap(d)
      
      assert pq.empty()
      for i in xrange(d ** randrange(2, 4) + randrange(d)):
        k = randrange(1000000)
        pq.put(k, k)
        #Test insert
        assert not pq.empty()
        assert pq.size() == i + 1
      
      print "---------------"
      A = pq.heapsort()
  
      #check that A is sorted (i.e. elements are popped from the queue in the right order
      B = sorted(A)
      assert(B == A)
    
    print "Decrease key in min heap"  
    #Check decrease_key
    for d in xrange(2,10):
      #Test d way
      print d
      pq = dway_heap(d)
      memo = {}
      assert pq.empty()
      for i in xrange(d ** randrange(2, 4) + randrange(d)):
        k = randrange(1000000)
        pq.put(k,k)
        memo[(k, k)] = 1
        #Test insert
        assert not pq.empty()
        assert pq.size() == i + 1
      memo = memo.keys()
      for _ in xrange(min(5, randrange(pq.size()))):
        index = randrange(len(memo))
        item = memo.pop(index)
        pq.decrease_priority(item[dway_heap.ELEMENT_INDEX], item[dway_heap.PRIORITY_INDEX] / 2)
        item = (item[dway_heap.PRIORITY_INDEX] / 2, item[dway_heap.ELEMENT_INDEX])
        memo.append(item)  
          
    print "Test max heap"
    for d in xrange(2,10):
      #Test d way
      print d
      pq = dway_heap(d, True)
      
      assert pq.empty()
      for i in xrange(d ** randrange(2, 4) + randrange(d)):
        k = randrange(1000000)
        pq.put(k, k)
        #Test insert
        assert not pq.empty()
        assert pq.size() == i + 1
      
      print "---------------"
      A = pq.heapsort()
  
      #check that A is sorted (i.e. elements are popped from the queue in the right order
      B = sorted(A, reverse=True)
      assert(B == A)
        
    print "Decrease key in max heap"  
    #Check decrease_key
    for d in xrange(2,10):
      #Test d way
      print d
      pq = dway_heap(d, True)
      memo = {}
      assert pq.empty()
      for i in xrange(d ** randrange(2, 4) + randrange(d)):
        k = randrange(1000000)
        pq.put(k,k)
        memo[(k, k)] = 1
        #Test insert
        assert not pq.empty()
        assert pq.size() == i + 1
      memo = memo.keys()
      for _ in xrange(min(5, randrange(pq.size()))):
        index = randrange(len(memo))
        item = memo.pop(index)
        pq.decrease_priority(item[dway_heap.ELEMENT_INDEX], item[dway_heap.PRIORITY_INDEX] * 2)
        item = (item[dway_heap.PRIORITY_INDEX] * 2, item[dway_heap.ELEMENT_INDEX])
        memo.append(item)      
      
    print "Test OK"   
  #END of test definition  
    
  test()

File: DWayHeap.java
import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedList;
import java.util.NoSuchElementException;

public class DWayHeap<T extends Comparable<T>> {

    /**
     * The actual container for elements added to the heap.
     */
    private LinkedList<T> heap;

    /**
     * States the branching factor of the heap - How many child each node has.
     */
    private int branch;
    
    /**
     * If true (by default) the heap is a minheap, otherwise a maxheap. 
     */
    public boolean minheap;

    /**
     * Constructor: creates a d-way minheap.
     * 
     * @param d The branching factor for the heap
     * @throws IllegalArgumentException Iff d < 2.
     */
    public DWayHeap(Integer d) throws IllegalArgumentException {
        init(d, true, null);
    }

    /**
     * Constructor: creates a d-way heap; whether it's a minheap or a maxheap depends
     * on the second parameter.
     * 
     * @param d The branching factor for the heap
     * @param minHeap If true the heap is a minheap (smallest element at its top);
     *                Otherwise it is a maxheap.
     * @throws IllegalArgumentException Iff d < 2.
     */    
    public DWayHeap(Integer d, boolean minHeap) throws IllegalArgumentException {
        init(d, minHeap, null);
    }

    /**
     * Constructor: creates a d-way minheap from the collection passed as second argumenr,
     * using the heapify algorithm to build it in linear time.
     * 
     * @param d The branching factor for the heap.
     * @param list The initial set of elements to be contained by the heap.
     * @throws IllegalArgumentException Iff d < 2.
     */    
    public DWayHeap(Integer d, Collection<T> list)
            throws IllegalArgumentException {
        init(d, true, list);
    }

    /**
     * Constructor: creates a d-way minheap from the collection passed as second argumenr,
     * using the heapify algorithm to build it in linear time; 
     * Whether it's a minheap or a maxheap depends on the second parameter.
     *  
     * @param d The branching factor for the heap.
     * @param minHeap If true the heap is a minheap (smallest element at its top);
     *                Otherwise it is a maxheap.
     * @param list The initial set of elements to be contained by the heap.
     * @throws IllegalArgumentException Iff d < 2.
     */    
    public DWayHeap(Integer d, boolean minHeap, Collection<T> list) {
        init(d, minHeap, list);
    }

    /**
     * Helper method to allow to reuse this class both as minheap and maxheap.
     * Evaluate which of the two elements passed as parameters is the smallest.
     * 
     * @param x First input to evaluate;
     * @param y Second input to evaluate;
     * @return true <=> x is smaller than y.
     */
    private boolean lt(T x, T y) {
        return minheap ? x.compareTo(y) < 0 : x.compareTo(y) > 0;
    }

    /**
     * Helper method to allow to reuse this class both as minheap and maxheap.
     * Evaluate which of the two elements passed as parameters is the smallest.
     * 
     * @param x First input to evaluate;
     * @param y Second input to evaluate;
     * @return true <=> x is not larger than y.
     */    
    private boolean le(T x, T y) {
        return lt(x, y) || x.compareTo(y) == 0;
    }

    /**
     * Helper method: initialize the heap.
     * @param d The branching factor for the heap.
     * @param minHeap If true the heap is a minheap (smallest element at its top);
     *                Otherwise it is a maxheap.
     * @param list The initial set of elements to be contained by the heap.
     * @throws IllegalArgumentException Iff d < 2.
     */
    private void init(Integer d, boolean minHeap, Collection<T> list)
            throws IllegalArgumentException {
        if (d < 2) {
            throw new IllegalArgumentException("Branching factor must be 2 or greater");
        }
        this.minheap = minHeap;
        this.branch = d;
        if (list != null && !list.isEmpty()) {
            heapify(list);
        } else {
            heap = new LinkedList<T>();
        }
    }

    /**
     * Given a position inside the heap, returns the index of its parent.
     * @param child The index of the element whose parent is required.
     * @return i >= 0 if child has a parent, -1 otherwise.
     */
    private int parentIndex(int child) {
        if (child == 0)
            return -1;
        else
            return (child - 1) / branch; // auto floor
    }

    /**
     * Given a position inside the heap, returns the index of its leftmost child.
     * This index might or might not be inside the heap - no check is performed at this stage.
     * @param parent The index of the element whose children must be located.
     * @return The index of the first child of the element passed.
     */
    private int childIndex(int parent) {
        return parent * branch + 1;
    }

    /**
     * Build a heap from a collection of elements, in linear worst case time.
     * Running time: O(n) in the worst case.
     * @param list The collection to be inserted in the heap.
     */
    private void heapify(Collection<T> list) {
        heap = new LinkedList<T>(list);
        
        int child, parent;
        T tmp;
        for (child = heap.size() - 1; child > 0; child--) {
            parent = this.parentIndex(child);
            if (this.lt(heap.get(child), heap.get(parent))) {
                tmp = heap.get(child);
                heap.set(child, heap.get(parent));
                heap.set(parent, tmp);
            }
        }
    }

    /**
     * Remove the top element from the heap, and returns it.
     * The top element is guaranteed to be the maximum in a maxheap, 
     * and the minimum in a minheap.
     * Running time: O(log_d(n)) in the worst case.
     * @return The top element in the heap, if it isn't empty.
     * @throws NoSuchElementException If the heap is empty.
     */
    public T remove() throws NoSuchElementException {
        T res = heap.getFirst(); // Stores the top element without removing it

        if (heap.size() == 1) {
            heap.removeFirst(); // Only one element
            return res;
        }
        T parentElem, bestChildElem;
        parentElem = heap.removeLast();

        int i, parent = 0, child = 1, bestChild;

        while (child < heap.size()) {

            bestChild = child;
            bestChildElem = heap.get(child);
            for (i = 1; i < branch && child + i < heap.size(); i++) {
                if (lt(heap.get(child + i), bestChildElem)) {
                    bestChild = child + i;
                    bestChildElem = heap.get(bestChild);
                }
            }
            if (lt(parentElem, bestChildElem)) {
                heap.set(parent, parentElem);
                return res; // The heap properties have been restored
            } else {
                heap.set(parent, bestChildElem);
                parent = bestChild;
                child = this.childIndex(parent);
            }
        }
        heap.set(parent, parentElem);
        return res;
    }

    /**
     * Add a new element to the heap, ensuring that the heap properties are not violated.
     * Running time: O(log_d(n)) in the worst case.
     * @param el The new element to be added.
     * @return This heap object, to allow method chaining.
     */
    public DWayHeap<T> add(T el) {
        heap.add(el);
        if (heap.size() > 1) {
            pullUp(heap.size() - 1);
        }
        return this;
    }
    
    /**
     * Helper method: climb an element towards the top of the heap until heap
     * properties are not violated anymore.
     * Running time: O(log(n)) in the worst case.
     * @param child The index of the element that need to be moved towards the top.
     */
    private void pullUp(int child) {
        T el = heap.get(child);
        
        int parent = this.parentIndex(child);
        T parentElem;
        while (parent >= 0) {
            parentElem = heap.get(parent);
            if (lt(el, parentElem)) {
                heap.set(child, parentElem);
                child = parent;
                parent = this.parentIndex(parent);
            } else {
                break;
            }
        }
        heap.set(child, el);        
    }
    
    /**
     * Decrease the priority of an element in the heap: it replaces one element of type T
     * with another one, which must be smaller [greater, for a maxheap]. 
     * Running time: O(log_d(n)) in the worst case.
     * @param oldElement The element to be replaced.
     * @param newElement The new element with whom the old one must be replaced. 
     *                   It must have a lower priority than the old one
     *                   (hence be smaller if it is a minheap, larger in a maxheap).
     * @return This heap object, to allow method chaining.
     * @throws NoSuchElementException If oldElement is not in the heap.
     * @throws IllegalArgumentException If the new element is greater [smaller, 
     *                                  for a maxheap] than the old one.
     */
    public DWayHeap<T> decreasePriority(T oldElement, T newElement) throws NoSuchElementException, IllegalArgumentException {
        if (lt(oldElement, newElement)) {
            throw new IllegalArgumentException();
        }
        int child = this.heap.indexOf(oldElement);
        
        if (child == -1) throw new NoSuchElementException();
        heap.set(child, newElement);
        pullUp(child);
        
        return this;
    }

    /**
     * Check that whether the heap contains at least one element.
     * @return True <=> the heap contains at least one element.
     */
    public boolean isEmpty() {
        return heap.isEmpty();
    }

    /**
     * Asserts the properties of the heap.
     */
    private void checkHeapProperties() {
        if (isEmpty()) {
            assert (heap.size() == 0);
        }

        int parent = 0, child = 1;

        while (child < heap.size()) {
            for (int i = 0; i < branch && child + i < heap.size(); i++) {
                try {
                    assert (le(heap.get(parent), heap.get(child + i)));
                } catch (AssertionError e) {
                    System.out.println(branch);
                    System.out.println(parent + " " + (child + i));
                    System.out.println(heap);
                    throw e;
                }
            }
            parent += 1;
            child = branch * parent + 1;
        }
    }
    
    
    /**
     * Tests the class
     */
    private static void test() {
        boolean exceptionThrown = false;
        
        //Branching factor must be >= 2
        try {
            new DWayHeap<Integer>(1);
        } catch (IllegalArgumentException e) {
            exceptionThrown = true;
        }
        assert (exceptionThrown);
        try {
            new DWayHeap<Integer>(-1);
        } catch (IllegalArgumentException e) {
            exceptionThrown = true;
        }
        assert (exceptionThrown);        

        //Test lt and le helper methods for minheaps and maxheaps
        DWayHeap<Integer> testHeap = new DWayHeap<Integer>(2);
        assert (testHeap.lt(1, 2));
        assert (testHeap.le(1, 1));
        assert (!testHeap.lt(2, 1));

        testHeap = new DWayHeap<Integer>(2, false);
        assert (testHeap.lt(2, 1));
        assert (testHeap.le(1, 1));
        assert (!testHeap.lt(1, 2));

        LinkedList<Integer> stack;

        // Min heap, branch factor d
        for (int d = 2; d < 7; d++) {
            stack = new LinkedList<Integer>();
            testHeap = new DWayHeap<Integer>(2);

            assert (testHeap.isEmpty());

            exceptionThrown = false;
            try {
                testHeap.remove();
            } catch (NoSuchElementException e) {
                exceptionThrown = true;
            }
            assert (exceptionThrown);

            testHeap.add((int) (Math.random() * 100));
            assert (!testHeap.isEmpty());
            for (int i = 0; i < 100 + (int) (50 * Math.random()); i++) {
                testHeap.add((int) (Math.random() * 100));
                
                testHeap.checkHeapProperties();
                if (Math.random() < 0.15) {
                    testHeap.remove();
                    testHeap.checkHeapProperties();
                }
            }

            stack.add(Integer.MIN_VALUE);

            while (!testHeap.isEmpty()) {
                int tmp = testHeap.remove();
                assert (tmp >= stack.getLast());
                stack.addLast(tmp);
            }
        }

        // Max heap, branch factor d
        for (int d = 2; d < 7; d++) {
            stack = new LinkedList<Integer>();
            testHeap = new DWayHeap<Integer>(2, false);

            assert (testHeap.isEmpty());

            exceptionThrown = false;
            try {
                testHeap.remove();
            } catch (NoSuchElementException e) {
                exceptionThrown = true;
            }
            assert (exceptionThrown);

            testHeap.add((int) (Math.random() * 100));
            assert (!testHeap.isEmpty());
            for (int i = 0; i < 100 + (int) (50 * Math.random()); i++) {
                testHeap.add((int) (Math.random() * 100));
                if (Math.random() < 0.15) {
                    testHeap.remove();
                    testHeap.checkHeapProperties();
                }
            }

            stack.add(Integer.MAX_VALUE);

            while (!testHeap.isEmpty()) {
                int tmp = testHeap.remove();
                try {
                    assert (tmp <= stack.getLast());
                } catch (AssertionError e) {
                    System.out.println(testHeap.heap);
                    throw e;
                }
                stack.addLast(tmp);
            }
        }
        // Test heapify
        DWayHeap<Double> testHeapDouble = new DWayHeap<Double>(3,
                Arrays.asList(new Double[] { 0.1, 4.0, 1.321, 3.1415, -7.1 }));
        testHeapDouble.checkHeapProperties();
        for (int i = 0; i < 50; i++) {
            testHeapDouble.add(Math.random());
        }
        testHeapDouble.checkHeapProperties();
        
        testHeapDouble = new DWayHeap<Double>(4, null);
        testHeapDouble.checkHeapProperties();
        for (int i = 0; i < 50; i++) {
            testHeapDouble.add(Math.random());
        }
        testHeapDouble.checkHeapProperties();
        
        testHeapDouble = new DWayHeap<Double>(3, false,
                Arrays.asList(new Double[] { 0.1, 4.0, 1.321, 3.1415, -7.1 }));
        testHeapDouble.checkHeapProperties();
        testHeapDouble.decreasePriority(4.0, 8.0);
        for (int i = 0; i < 50; i++) {
            testHeapDouble.add(Math.random());
        }
        testHeapDouble.checkHeapProperties();
        
        testHeap = new DWayHeap<Integer>(4);
        
        for (Integer i : new Integer[]{3, 1, 5, 2, 4, 6, 8, 7, 0}){
            testHeap.add(i);
        }
        
        int j = 0;
        for (int i = 0; i < 9; i++) {
            try {
                j = testHeap.remove();
                assert(j == i);
            } catch(AssertionError e) {
                System.out.println(i + " " + j);
                System.out.println(testHeap.heap);
                throw e;
            }
        }
    }

    /**
     * @param args
     */
    public static void main(String[] args) {
        test();
    }

}
